flutter-for-ios.md

```diff
@@ -27,3 +27,1408 @@ that are most relevant to your needs.
+}
+
+class SampleApp extends StatelessWidget {
+  // This widget is the root of your application.
+  @override
+  Widget build(BuildContext context) {
+    return new MaterialApp(
+      title: 'Sample App',
+      theme: new ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: new SampleAppPage(),
+    );
+  }
+}
+
+class SampleAppPage extends StatefulWidget {
+  SampleAppPage({Key key}) : super(key: key);
+
+  @override
+  _SampleAppPageState createState() => new _SampleAppPageState();
+}
+
+class _SampleAppPageState extends State<SampleAppPage> {
+  @override
+  Widget build(BuildContext context) {
+    return new Scaffold(
+      appBar: new AppBar(
+        title: new Text("Sample App"),
+      ),
+      body: new ListView(children: _getListData()),
+    );
+  }
+
+  _getListData() {
+    List<Widget> widgets = [];
+    for (int i = 0; i < 100; i++) {
+      widgets.add(new Padding(padding: new EdgeInsets.all(10.0), child: new Text("Row $i")));
+    }
+    return widgets;
+  }
+}
+{% endprettify %}
+
+## How do I know which list item is clicked?
+
+In iOS, you would implement the delegate method `tableView:didSelectRowAtIndexPath:`. Flutter makes it even easier by letting you just use the touch handling that the widgets you passed in have.
+
+<!-- skip -->
+{% prettify dart %}
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(new SampleApp());
+}
+
+class SampleApp extends StatelessWidget {
+  // This widget is the root of your application.
+  @override
+  Widget build(BuildContext context) {
+    return new MaterialApp(
+      title: 'Sample App',
+      theme: new ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: new SampleAppPage(),
+    );
+  }
+}
+
+class SampleAppPage extends StatefulWidget {
+  SampleAppPage({Key key}) : super(key: key);
+
+  @override
+  _SampleAppPageState createState() => new _SampleAppPageState();
+}
+
+class _SampleAppPageState extends State<SampleAppPage> {
+  @override
+  Widget build(BuildContext context) {
+    return new Scaffold(
+      appBar: new AppBar(
+        title: new Text("Sample App"),
+      ),
+      body: new ListView(children: _getListData()),
+    );
+  }
+
+  _getListData() {
+    List<Widget> widgets = [];
+    for (int i = 0; i < 100; i++) {
+      widgets.add(new GestureDetector(
+        child: new Padding(
+          padding: new EdgeInsets.all(10.0),
+          child: new Text("Row $i"),
+        ),
+        onTap: () {
+          print('row tapped');
+        },
+      ));
+    }
+    return widgets;
+  }
+}
+{% endprettify %}
+
+## How do I update `ListView`s dynamically?
+
+In iOS, you would update the data for the list view, and notify the table or
+collection view using the `reloadData` method. In Flutter if you were to update
+the list of widgets inside a `setState()`, you would quickly see that your
+data did not change visually.
+
+This is because when `setState()` is called, the Flutter rendering engine will
```

Comment:
Unsure about this; needs verification

Reply:
Summoning @takhion the wise

Reply:
###### _...flash of light, puff of smoke, and magical sounds!_

I believe there might be a fundamental misconception at play here: [`State.setState`] **is** the mechanism though which [`StatefulWidget`]s notify changes, and as a direct consequence the [`State.build`] function is executed again, returning a (potentially) different sub-tree of [`Widget`]s.

A new [`Widget`] instance means nothing by itself, since it's up to the corresponding [`Element`] (aka the [`BuildContext`]) to decide whether the new instance is of the same type and carries new data (if you're curious, check out [`Element.updateChild`] and [`Widget.canUpdate`]).

Now, this specific example is _misleading_ since apparently the [`ListView`] doesn't draw new rows unless the `widgets` list is re-created... But that has nothing to do with the usual way widgets work!

The issue here is that this particular "configuration" of [`ListView`] is really optimised for having a static/immutable list. As such, internally it creates [a specific `childrenDelegate`][`SliverChildListDelegate.init`] ([`SliverChildListDelegate`]) that [when prompted whether to "rebuild"][`SliverChildListDelegate.shouldRebuild`] assumes that the lists are immutable, so it simply compares the old and the new list instances with the equality operator, which for `List` is the default check by reference. In conclusion, by creating our own delegate and forcing `shouldRebuild` to return `true`, we can treat the list as mutable and everything works as expected!

🎉🎉🎉

<details>
<summary><b>example</b></summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(new SampleApp());
}

class SampleApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Sample App',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new SampleAppPage(),
    );
  }
}

class SampleAppPage extends StatefulWidget {
  SampleAppPage({Key key}) : super(key: key);

  @override
  _SampleAppPageState createState() => new _SampleAppPageState();
}

class _SampleAppPageState extends State {
  // [List] needs a type to make Dart 2 happy
  List<Widget> widgets = [];

  @override
  void initState() {
    super.initState();
    for (int i = 0; i < 10; i++) {
      widgets.add(getRow(i));
    }
  }

  @override
  Widget build(BuildContext context) {
    //// this doesn't work
    //var listView = new ListView(children: widgets);

    // this works!
    var listView = new ListView.custom(
      childrenDelegate: new _MyListDelegate(widgets),
    );

    print("build!\n"
        // context is always the same instance
        "* context=${identityHashCode(context)}\n"
        // listView is always a different instance
        "* listView=${identityHashCode(listView)}");

    return new Scaffold(
      appBar: new AppBar(
        title: new Text("Sample App"),
      ),
      body: listView,
    );
  }

  Widget getRow(int i) {
    return new GestureDetector(
      child: new Padding(
        padding: new EdgeInsets.all(10.0),
        child: new Text("Row $i"),
      ),
      onTap: () {
        setState(() {
          widgets.add(getRow(widgets.length + 1));
          print('row $i (widgets.length=${widgets.length})');
        });
      },
    );
  }
}

class _MyListDelegate extends SliverChildListDelegate {
  _MyListDelegate(List<Widget> children) : super(children);

  @override
  bool shouldRebuild(SliverChildListDelegate oldDelegate) => true;
}
```

</details>

[`BuildContext`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L1748-L1811
[`Element`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L2425-L2475
[`Element.updateChild`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L2637-L2669

[`Widget`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L338-L373
[`Widget.canUpdate`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L422-L435

[`StatefulWidget`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L592-L773
[`State.setState`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L1042-L1086
[`State.build`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/framework.dart#L1191-L1291

[`ListView`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/scroll_view.dart#L433-L570

[`SliverChildDelegate.shouldRebuild`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/sliver.dart#L86-L95

[`SliverChildListDelegate`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/sliver.dart#L214-L243
[`SliverChildListDelegate.shouldRebuild`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/sliver.dart#L302-L306
[`SliverChildListDelegate.init`]: https://github.com/flutter/flutter/blob/v0.3.5/packages/flutter/lib/src/widgets/scroll_view.dart#L599-L603

Reply:
Thanks Eugy, I'll rephrase this section to match :) ❤️